### PR TITLE
WT-6403 restore format non-timestamp transactional testing 

### PR DESCRIPTION
**Pull request by keithbostic**
Restore non-timestamp transaction testing in format. Format now runs in one of two modes, either timestamp (TS) transactions or non-timestamp (non-TS) transactions, governed by the transaction.timestamp configuration string, turning that configuration string off configures non-TS runs.

TS transactions only use snapshot isolation, non-TS transactions do all updates in snapshot isolation, but do reads in both snapshot and lower-level isolations.

WT_CURSOR.remove inside an implicit transaction could fail with "key not set" if the underlying btree routine returned rollback.  The current behavior is we must fail if the cursor position is initially set and we lose that position somehow, change the API's implicit transaction macros to return WT_ROLLBACK in the case where retry isn't an option.
    Remove the TXN_API_END_RETRY macro, it has only one caller, rename TXN_API_END.
    Remove a test in TXN_API_END() of ret == 0, it cannot be 0 at that point.

Update method documentation to note that all object updates and calls to WT_SESSION.query_timestamp must be done at snapshot isolation.

Remove documentation note that all uncommitted transactions must fit into memory, that is no longer true (for example, prepared transactions).

Update format to not configure the isolation level where the default level (snapshot) is correct.

Remove disabled lower-level isolation test from evergreen.yml, TS runs are never going to support isolation levels other than snapshot.

Remove format's transaction.isolation configuration string, TS runs will never support isolation levels other than snapshot, and non-TS runs will only do reads at lower-levels than snapshot.

Rename format's "transaction.frequency" configuration string to "transation.implicit", as TS runs do not support implicit transactions. The new configuration string sets the percentage of non-TS operations that are done inside implicit transactions, rather than explicit transactions.

Remove the assert.commit_timestamp string, it's no longer supported by the library.

Add an assert.read_timestamp configuration string to support "assert=(read_timestamp=VALUE)", with the value of always in the case of TS runs, and never in the case of non-TS runs.

Add an assert.write_timestamp configuration string to support the library configuration "assert=(write_timestamp=on), write_timestamp_usage=VALUE", with the value of always in the case of TS runs, and never in the case of non-TS runs.

Remove the transaction.rollback_to_stable configuration string and change it to always be called in both TS and non-TS runs. Change rollback-to-stable to not do additional checks against saved operations unless in a TS run.

### CODE REVIEW

**Comment by keithbostic** [comment]
@lukech, this restores format testing of non-timestamp transactions, including lower-level isolation and implicit transactions (but only in non-timestamp transactions). Additionally, rollback-to-stable is no executed on all on-disk databases. This is notable because rollback-to-stable was turned off entirely in format until this change.

Do you want to run this branch separately in a stress-testing environment before we merge it into develop?

**Comment by keithbostic** [comment]
@agorrod, you might want to skim these changes and make sure I haven't violated any of the rules.

**Review by ddanderson — APPROVED** [review]
LGTM

**Review by kommiharibabu — APPROVED** [review]
LGTM. Only doubt of RTS change for non-ts transactions.

File: test/format/ops.c [hunk 1]
```diff
@@ -201,28 +201,37 @@ tinfo_teardown(void)
...
+    /* Rollback-to-stable only makes sense for on-disk stores. */
+    if (g.c_in_memory != 0)
+        return;
+
     testutil_check(__wt_snprintf(cmd, sizeof(cmd), ROLLBACK_STABLE_COPY_CMD, g.home));
     if ((ret = system(cmd)) != 0)
         testutil_die(ret, "rollback to stable copy (\"%s\") failed", cmd);
     trace_msg("%-10s ts=%" PRIu64, "rts", g.stable_timestamp);
 
     g.wts_conn->rollback_to_stable(g.wts_conn, NULL);
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply test/format/ops.c]: @kommiharibabu, this was my mistake, thank you for catching it, I didn't understand how RTS works with recovery.

Should RTS fail in the non-timestamp case?

The WT_CONNECTION.rollback_to_stable documentation says an application must set the stable timestamp, but that isn't triggering an error as far as I can tell. Presumably, recovery is setting a stable timestamp and so we're not noticing the application is doing something wrong? Is there a way we could detect this API mistake?

I've pushed a fix for this problem in 9b0ed5002.

**Comment by keithbostic** [comment]
@kommiharibabu, based on stress failures (found here: https://evergreen.mongodb.com/task_log_raw/wiredtiger_ubuntu1804_ppc_format_stress_sanitizer_smoke_ppc_test_patch_cf79d062ecdaa7509ca80c4a643268ee5303e915_608f971a2fbabe71f8ac5eae_21_05_03_06_25_24/0?type=T&text=true), I've committed a set of additional changes, mostly in areas you're familiar with, I believe.

Would you please do the review? It includes the following commits: 936debd76, 1622f9eab and 3e8a6f38e.

**Comment by keithbostic** [comment]
@kommiharibabu, in this stress test output:

https://evergreen.mongodb.com/task_log_raw/wiredtiger_ubuntu1804_ppc_format_stress_smoke_test_2_patch_cf79d062ecdaa7509ca80c4a643268ee5303e915_608f971a2fbabe71f8ac5eae_21_05_03_06_25_24/0?type=T&text=true

There's an exciting failure:
```
[2021/05/03 11:00:28.826]     0x1003ac16cb0:oldest pinned transaction ID rolled back for eviction
[2021/05/03 11:00:28.826]     0x1003ac18450:oldest pinned transaction ID rolled back for eviction
[2021/05/03 11:00:28.826]     [1620039623:962768][34245:0x7a5c00d96f50], file:wt.wt, WT_SESSION.checkpoint: __wt_block_read_off, 327: wt.wt: read checksum error for 17408B block at offset 6712832: block header checksum of 0x726f6d65 doesn't match expected checksum of 0xd779cbd1
```

I'm wondering how safe it is to roll back a pinned transaction ID in the middle of rollback-to-stable?

**Comment by keithbostic** [comment]
@kommiharibabu, @lukech, not sure if you are both aware: we weren't doing any RTS calls from format before this branch, so it isn't too surprising that we're seeing RTS failures in these runs.

File: test/format/ops.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply test/format/ops.c]: Thanks. RTS was failing for non-timestamp case when there is no stable timestamp is set. But as part of supporting RTS for non-timestamp tables, we remove that restriction. If the rollback to stable is called on a running system where non-timestamp tables are present, it just doesn't do anything.

**Comment by kommiharibabu** [comment]
@keithbostic There shouldn't be any active transactions that are running when the RTS is operating. RTS checks for any active transactions before it starts, but if any transaction starts later before it finishes, it can cause problems. I am not sure whether the above-mentioned transaction is the reason for the failure.

File: test/evergreen.yml [hunk 2]
```diff
@@ -2061,10 +2061,6 @@ tasks:
       - func: "format test"
         vars:
           extra_args: checkpoints=1 leak_memory=0 mmap=1 file_type=row compression=zlib huffman_value=1
-      # FIXME-WT-6668: temporarily disable lower isolation level test
-      # - func: "format test"
-      #   vars:
-      #     extra_args: checkpoints=1 leak_memory=0 mmap=1 file_type=row isolation=random transaction_timestamps=0
```
Comment (lukech) [inline test/evergreen.yml:2067]: Should we retain/re-enable this test to cover non-timestamp (standalone wiredtiger) run with lower isolation levels (read)?

Comment (keithbostic) [reply test/evergreen.yml:2067]: @lukech, this test no longer makes sense -- when `timestamps=0` is set, lower-level isolation is included in all cases, not just when specifically configured (and the `isolation` configuration has been removed).

Comment (lukech) [reply test/evergreen.yml:2067]: Alright, thanks for the explanation - it makes sense to me now. 
Just hope to make sure we still have coverage for lower isolation levels.

**Review by lukech — APPROVED** [review]

**Review by quchenhao — APPROVED** [review]
Sorry, I missed this one.

**Comment by keithbostic** [comment]
New stress run: https://evergreen.mongodb.com/version/60950afa0ae60671c07609ec?redirect_spruce_users=true

**Comment by keithbostic** [comment]
Stress: https://spruce.mongodb.com/version/609af8dcc9ec44376e6f67ef/tasks

**Comment by kommiharibabu** [comment]
@keithbostic I reviewed the WT_ROLLBACK fix changes and they are looks good to me.

**Comment by ddanderson** [comment]
@keithbostic I've reviewed the two commits you flagged, they look good to me.

**Comment by keithbostic** [comment]
stress: https://evergreen.mongodb.com/version/609bf41e3627e0686eb7a94d?redirect_spruce_users=true

**Comment by keithbostic** [comment]
@sueloverso, hopefully you remember how format's configuration works?

Would you be please review 36d66dd (well, basically, the whole `config_transaction()` function)? I think it's self-explanatory, but if it isn't don't waste time, I'm happy to walk you through it.

**Review by sueloverso — APPROVED** [review]
I reviewed the two commits you noted. LGTM. They make sense to me and I like the amount of removal!

**Comment by keithbostic** [comment]
stress: https://evergreen.mongodb.com/version/609d4c49c9ec444f520966ed?redirect_spruce_users=true

**Pull request merged**